### ISSUE TITLE
Consider the impact of ECS attribute alignment on all existing tracing semantic conventions

### ISSUE THREAD

**Issue by trask** [open-telemetry/opentelemetry-specification#3408]
## [HTTP](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md)
- [ ] #3181
- [ ] #3182
- [ ] #3199 
- [ ] #3371 

## [Exceptions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/exceptions.md)
- [ ] #3198 

## [General remote service attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/span-general.md#general-remote-service-attributes)

@AlexanderWert to confirm that ECS doesn't have an equivalent to `peer.service`.

## [General identity attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/span-general.md#general-identity-attributes)

| OpenTelemetry | ECS |
| --- | --- |
| `enduser.id` | `user.id` |
| `enduser.role`, `enduser.scope` | `user.roles` (?) |

## [General thread attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/span-general.md#general-thread-attributes)

| OpenTelemetry | ECS |
| --- | --- |
| `thread.id` | `process.thread.id` |
| `thread.name` | `process.thread.name` |

## [Source code attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/span-general.md#source-code-attributes)

| OpenTelemetry | ECS |
| --- | --- |
| `code.function` | `log.origin.function` |
| `code.namespace` |  |
| `code.filepath` | `log.origin.file.name` |
| `code.lineno` | `log.origin.file.line` |
| `code.column` |   |

## [Messaging](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md)

This is also discussed/tracked at #3196.

Only impacted by network attribute changes above (#3199 and #3371).

(ECS doesn't have any messaging-specific attributes, @AlexanderWert to confirm)

Could also be impacted by conflict with `source.*` and `destination.*` namespace, see https://github.com/open-telemetry/opentelemetry-specification/issues/3407

## [Database](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md)

Only impacted by network attribute changes above (#3199 and #3371).

(ECS doesn't have any database-specific attributes, @AlexanderWert to confirm).

## [RPC](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/rpc.md)

Only impacted by network attribute changes above (#3199 and #3371).

(ECS doesn't have any RPC-specific attributes, @AlexanderWert to confirm).

## [FaaS](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/faas.md)

TBD, see https://www.elastic.co/guide/en/ecs/current/ecs-faas.html

## [CloudEvents](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/cloudevents.md)

ECS doesn't have any cloud event specific attributes (@AlexanderWert to confirm)

## [Feature Flags](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/feature-flags.md)

ECS doesn't have any feature flag specific attributes (@AlexanderWert to confirm)

**Comment by trask** [comment]
@Oberon00 thx for calling out the ECS alignment issue in #3188, sorry I missed that

@cartersocha @tylerbenson we do intend to factor in ECS alignment across all semantic conventions (https://github.com/open-telemetry/oteps/pull/222)

where there are conflicts, there's no blanket rule about which side (or mix) we will adopt going forward, and so we need to analyze these conflicts on a case-by-case basis, factoring in a few things (among potentially others):
* are there inherent advantages to one modeling or the other?
* is the naming clearer on one side or the other?
* how widespread is adoption of both sides?

this is what we need to gather in this issue so that we can make these case-by-case decisions

**Comment by tsloughter** [comment]
I think the naming is clearer on our side. These were very recent changes because we felt the naming was clearer.

Also, I don't think `faas.trigger.request_id` is the same as our invocation_id, but I'm not really sure what it is. It sounds like it'd be like if the trigger is HTTP and that contains a request id header.

**Comment by Oberon00** [comment]
HTTP is not mentioned. Note that in https://github.com/open-telemetry/opentelemetry-specification/pull/3378 it was noted that our faas.invocation_id is not suitable as (job) execution_id, I imagine this might have been different with the old name. So maybe there is indeed space for both an execution(_id) and an invocation_id/request ID

**Comment by cartersocha** [comment]
@trask the OTel version is slightly more up to date and vendor neutral (although still Lambda skewed). @Oberon00 called out most of the changes already and presciently mentioned ECS previously. 

I believe ECS is missing the faas.max_memory field and ECS uses a lambda specific faas.id like we used to. Personally I think there's also a lot of value in having a unique cloud.resource_id that is consistent for all cloud resources. The AWS ARN is already that but it's currently restricted to just being a faas specific field.

**Comment by Oberon00** [comment]
> uses a lambda specific faas.id like we used to

This is not lambda specific, it has definitions for Azure and Google Cloud Functions as well

**Comment by tsloughter** [comment]
@Oberon00 HTTP is one of the trigger types.

**Comment by Oberon00** [comment]
@tsloughter There is a separate http.request.id. https://www.elastic.co/guide/en/ecs/current/ecs-http.html#field-http-request-id

**Comment by tsloughter** [comment]
@Oberon00 right, seems redundant. Unless it has some other meaning, it doesn't make it clear and the example is just a number.

**Comment by AlexanderWert** [comment]
> Also, I don't think faas.trigger.request_id is the same as our invocation_id, but I'm not really sure what it is. It sounds like it'd be like if the trigger is HTTP and that contains a request id header.

The intention of `faas.trigger.request_id` was to capture an identifier of the trigger event. For HTTP (e.g. through Api Gateway on AWS in case of lambda) it's the API Gateway request ID, in case of S3 events it's the `x-amz-request-id` (see here: https://docs.aws.amazon.com/lambda/latest/dg/with-s3.html). but I admit that it mostly makes sense with HTTP, since with other events there are often batches of events triggering a function invocation. So it's not the same as `faas.invocation_id` in OTel SemConv. 

In general, the `faas.*` fields in ECS were introduced based on and in alignment with the existing OTel `faas.*` attributes at that time (only exception was the `faas.trigger.type` and `faas.trigger.request_id`). Also, `faas.*` fields in ECS are still in Beta, so it's an area where it makes more sense to align with existing OTel attributes.

If we adopt `http.request.id` from ECS, that could be a good way to capture the above (`faas.trigger.request_id`).

**Comment by AlexanderWert** [comment]
> General remote service attributes
> @AlexanderWert to confirm that ECS doesn't have an equivalent to peer.service.

ECS `service.*` fields [can be self-nested ](https://www.elastic.co/guide/en/ecs/current/ecs-service.html#ecs-service-nestings) under `service.origin.*` and `service.target.*`, which has a similar purpose as the `peer.service` in OTel. Similar to our previous discussions on `client / server` and `source / destination`, here `service.origin.*` and `service.target.*` make the relationship more explicit, without the need to know the context of the field.

> Messaging
> (ECS doesn't have any messaging-specific attributes, @AlexanderWert to confirm)

:+1: 

> # RPC
> Only impacted by network attribute changes above (https://github.com/open-telemetry/opentelemetry-specification/issues/3199 and https://github.com/open-telemetry/opentelemetry-specification/issues/3371).
> 
> (ECS doesn't have any RPC-specific attributes, @AlexanderWert to confirm).

:+1: 

> # Database
> Only impacted by network attribute changes above (https://github.com/open-telemetry/opentelemetry-specification/issues/3199 and https://github.com/open-telemetry/opentelemetry-specification/issues/3371).
>
> (ECS doesn't have any database-specific attributes, @AlexanderWert to confirm).

:+1: 

> ECS doesn't have any cloud event specific attributes (@AlexanderWert to confirm)

That's correct that there are no cloud events fields in ECS. Though, there is a more [generic field set](https://www.elastic.co/guide/en/ecs/current/ecs-event.html) for `event.*`. So, there is some potential overlap here:

| OTel | ECS |
| --- | ---|
| `cloudevents.event_id` | `event.id` |
| `cloudevents.event_source` | `event.provider` |
| `cloudevents.event_spec_version ` |  |
| `cloudevents.event_type ` | `event.type` |
| `cloudevents.event_subject` |  |

> ECS doesn't have any feature flag specific attributes (@AlexanderWert to confirm)

:+1:

**Comment by svrnm** [comment]
Is this done?

**Comment by trask** [comment]
> Is this done?

I suspect it can be closed. I'll transfer it to semconv repo and let maintainers there decide.
